static/game/bullet.js

```diff
@@ -131,3 +144,137 @@ export class Bullet extends Entity {
+   */
+  constructor(
+    pos = new Vector(0, 0),
+    dir = new Vector(0, 0),
+    acc = new Vector(0, 0),
+    owner,
+    color = "white",
+    lifetime = 100,
+    damage = 1
+  ) {
+    super(
+      pos,
+      new Vector(0, 0),
+      new Vector(0, 0),
+      owner,
+      color,
+      lifetime,
+      damage
+    );
+    this.length = 0;
+    this.dir = dir.norm2();
+    this.lifetime = 100;
+    this.occludedByWalls = false;
+    // beams deal damage 4 times per second
+    this.maxCounter = 25;
+    this.counter = this.maxCounter - 1;
+  }
+
+  /**
+   * @override
+   */
+  getCollisionShape() {
+    return new CollisionBeam(
+      this.pos,
+      this.pos.add(this.dir.mult(this.length)),
+      this.width
+    );
+  }
+
+  /**
+   * @override
+   */
+  draw() {
+    this.drawPos = this.owner.drawPos.add(
+      this.owner.facing.mult(Math.min(this.owner.width) / 4)
+    );
+    line(
+      this.drawPos,
+      this.pos.add(this.dir.mult(this.length)),
+      this.color,
+      this.width
+    );
+    line(
+      this.drawPos,
+      this.pos.add(this.dir.mult(this.length)),
+      "white",
+      Math.max(this.width - 16, 0)
+    );
+  }
+
+  /**
+   * called when this beam touches an enemy (for enemy beams this is called
+   * when it touches the hero)
+   * @param {import("./creature.js").Creature} creature
+   */
+  touchEnemy(creature) {
+    // rate limit dealing damage to creatures
+    if (this.counter === this.maxCounter) {
+      // deal basic damage
+      creature.takeDamage(this.damage, this.dir);
+      // impart momentum
+      const size = (creature.width * creature.height) / 300;
+      creature.vel = creature.vel.add(this.dir.mult(this.knockback / size));
+      // call onHitEnemy functions
+      for (const ohe of this.onHitEnemy) {
+        if (ohe.func) ohe.func(this, ohe.data, creature);
+      }
+    }
+  }
+
+  destroy() {
+    // execute all on-destroy functions
+    for (const od of this.onDestroy) {
+      if (od["func"]) od["func"](this, od["data"]);
+    }
+  }
+
+  /**
+   * what to do when hitting a block
+   * @override
+   * @param {Vector} pos
+   */
+  collideWithBlock(pos) {
+    const cellVec = getCell(pos);
+    if (
+      inbounds(cellVec.x, cellVec.y) &&
+      blockField[cellVec.x][cellVec.y].durability !== Infinity
+    ) {
+      destroyBlock(cellVec, this.type === "PlayerBullet");
+    }
+  }
+
+  /**
+   * @override
+   * calculate length and set position each step
+   */
+  action() {
+    if (this.owner.deleteMe) this.deleteMe = true;
+    this.dir = this.owner.facing.rotate(this.angle);
+    this.pos = this.owner.pos.add(
+      this.owner.facing.mult(Math.min(this.owner.width) / 4)
+    );
+    const intersect = nextIntersection(this.pos, this.dir);
+    // rate limit block collisions
+    if (this.counter === this.maxCounter)
+      this.collideWithBlock(intersect);
+    this.length = intersect.sub(this.pos).mag();
+
+    if (this.counter++ > this.maxCounter) this.counter = 0;
```

Comment:
This should be called something more descriptive than "counter"
Also, this means that it can only damage one creature per beam, as they all share a single counter, which seems wrong.

Reply:
"DamageCounter"
"DamageTickCounter"
"DamageCooldown" (and make it go to 0)

Reply:
The problem I need to solve is that you don't want the beam dealing damage every single game step or it'll just nuke creatures. I'm starting to dislike this behavior anyway, since if you're lucky you can slip through a beam without getting hit. I'm thinking instead of having a map of creatures that have already been hit and making it wait an amount of time before hitting the same creature again.

Reply:
yeah that was gonna be my recommended solution

Reply:
Why don't we just do damage every tick but do like a 10th the bullet damage?

Reply:
You can't apply one tenth of the `onHitEnemy` functions though

Reply:
I implemented a smarter behavior using a map of creatures hit, each with their own cooldown in 8680f05

Reply:
nice